finagle-core/src/main/scala/com/twitter/finagle/tracing/Trace.scala

```diff
@@ -69,14 +86,15 @@ object Trace {
       val parent64 = ByteArrays.get64be(bytes, 8)
       val trace64 = ByteArrays.get64be(bytes, 16)
       val flags64 = ByteArrays.get64be(bytes, 24)
+      val traceIdHigh = if(body.length == 40) ByteArrays.get64be(bytes, 32) else 0L
 
       val flags = Flags(flags64)
       val sampled = if (flags.isFlagSet(Flags.SamplingKnown)) {
         if (flags.isFlagSet(Flags.Sampled)) someTrue else someFalse
       } else None
 
       val traceId = TraceId(
-        if (trace64 == parent64) None else Some(SpanId(trace64)),
+        if (trace64 == parent64) None else Some(SpanId(traceIdHigh, trace64)),
```

Comment:
For compat maybe better to add another spanid for high bits vs expand spanid to 128